Test Completed

homeindexjeff opened the pull request:
Both test-one and test-two have been completed as requested for the Junior Developer Role.
- Jeffrey F

danwfeeney commented on test-one/package.json:
```diff
@@ -7,7 +7,7 @@
     "test": "./node_modules/.bin/_mocha \"lib/**/*-tests.js\"",
     "go": "node lib/index",
-    "start": "nodemon -x \"npm run compile && npm run test && npm go\""
+    "start": "nodemon -x \"npm run compile && npm run test && npm run go\""
```
👍  good spot, fixed this up in master now

danwfeeney commented on test-one/src/index.js:
```diff
@@ -2,10 +2,22 @@ import Logger from './helpers/logger';
-
-logger.log(welcomeMessage);
\ No newline at end of file
+for(var i = 1; i<=100; i++){
```
should this be in the main JS file? how would we write unit tests for this?

danwfeeney replied:
should we have unit tests around this functionality?

dazMurph90 replied:
Possibly not, I have now moved this into its own class as to make testing easier.

dazMurph90 replied:
I have now added testing for the BossHog class.

danwfeeney commented on test-two/src/Controller.js:
```diff
@@ -0,0 +1,142 @@
+    
+    //retreives all usernames from the github API
+    //runs getUsersByPage by amountPages times
+    getAllUsers(callback){
```
Could this have been achieved with promises instead of callbacks to be consistent with the existing codebase?

dazMurph90 replied:
Old habits die hard. Fixed.

danwfeeney commented on test-two/src/index.js:
```diff
@@ -1,20 +1,57 @@
 const baseUrl = "https://api.github.com/";
 const organisationId = "uber";
-const gitHubApiAuthToken = ""; //add your GitHub API OAuth key here to increase request limit
+const gitHubApiAuthToken = "915fa9458cabcffab770ba07faadfc0b23ea707e"; //OAuth key
```
is there a way we could store this without exposing our token to the GitHub world?

dazMurph90 replied:
Good point. I have moved the key into a file external of the source code.

danwfeeney commented on test-two/src/index.js:
```diff
@@ -1,20 +1,57 @@
+
+//github API paginates the list of ubers users to 30 per page
+//we need to get amount of pages first(IE amount of requests to make)
+gitHubService.getHeaderForUsers(organisationId).then(function(header){
```
is there a way we could avoid making this initial request?

danwfeeney replied:
could this function definition use ES6 arrow function syntax for consistency with the existing code base?

dazMurph90 replied:
Sure. Patched up now.

danwfeeney commented on test-two/src/index.js:
```diff
@@ -1,20 +1,57 @@
+/*  -------------------------- MISC FUNCTIONS -------------------------------  */
+
+//extracts amount of pages for users from the API response header
+function getAmountOfPagesFromHeader(header){
```
does this functionality belong in the main program file?

danwfeeney replied:
Should this use ES6 arrow function syntax

danwfeeney replied:
Does this need to be unit tested?

dazMurph90 replied:
Perhaps not. I've moved this into the GitHubService class.

dazMurph90 replied:
Yes. This has now moved to a class for reuse, but should have been:
`var getAmountOfPagesFromHeader = (header) => { ... }`

dazMurph90 replied:
Unit test added.

danwfeeney commented on test-two/src/index.js:
```diff
@@ -1,20 +1,57 @@
+gitHubService.getHeaderForUsers(organisationId).then(function(header){
+    //param header returns the header information from an API call
+    
+    controller.amountPages = getAmountOfPagesFromHeader(header)
```
👍  for using the header to determine number of pages

danwfeeney commented on test-two/src/services/github.js:
```diff
@@ -5,13 +5,24 @@ class GitHubService {
 
-    getUsersForOrganisation(organisationId) {
-        return this.http.get(this.baseUrl + "orgs/" + organisationId + "/members" + this.authenticationQueryString);
+    getUsersForOrganisation(organisationId,pageNumber) {
```
👍  for maintaining a reusable class

danwfeeney commented on test-two/src/services/github.js:
```diff
@@ -5,13 +5,24 @@ class GitHubService {
+        return this.http.getHeader(this.baseUrl + "orgs/" + organisationId + "/members?" + this.authenticationQueryString);
+    }
+    
+    getUsersRepos(username){
```
What happens if a user has more than 100 repos? Is 100 too many repos to return in some cases?

dazMurph90 replied:
Ah, that is an oversight.  will get this fixed in an update soon.

danwfeeney commented on test-two/src/services/github.js:
```diff
@@ -5,13 +5,24 @@ class GitHubService {
-    getUsersForOrganisation(organisationId) {
-        return this.http.get(this.baseUrl + "orgs/" + organisationId + "/members" + this.authenticationQueryString);
+    getUsersForOrganisation(organisationId,pageNumber) {
+        return this.http.get(this.baseUrl + "orgs/" + organisationId + "/members?page=" + pageNumber + "&" + this.authenticationQueryString);
```
Is there a cleaner way to concatenate these strings in ES6?

dazMurph90 replied:
I've updated this to use template string substitutions.

homeindexjeff committed:
Test Completed

Both test-one and test-two have been completed as requested for the Junior Developer Role.
- Jeffrey F

danwfeeney commented on test-one/.vs/ProjectSettings.json:
```diff
@@ -0,0 +1,3 @@
+{
```
Should this folder be committed to source control?

homeindexjeff replied:
I'm not to sure about that. I never used that file and didn't know it was there until showing the hidden folders in Visual Studio.

danwfeeney commented on test-one/src/index-test.js:
```diff
@@ -9,5 +9,5 @@ test("logged welcome message", async () => {
   await go();
 
-  expect(logger.log).toBeCalledWith(expectedOutput);
+  //expect(logger.log).toBeCalledWith(expectedOutput);
```
Why has this been commented out?

homeindexjeff replied:
Sorry was in a little bit of a rush when I did this index-test.js should of looked like this:

import { go } from "./index";
import * as logger from "./helpers/logger";

test("Confirmation Test", async () => {
    var counter;
    var displayMessage = "";

    for (counter = 1; counter <= 100; counter++) {
        if (counter % 3 && counter % 5) {
            displayMessage += counter;
        } else {
            if (counter % 3 == 0) {
                displayMessage += "Boss";
            }
            if (counter % 5 == 0) {
                displayMessage += "Hog";
            }
        }
        displayMessage += "\n";
    }

  logger.log = jest.fn();

  await go();

  expect(logger.log).toBeCalledWith(displayMessage);
});

danwfeeney commented on test-one/src/index.js:
```diff
@@ -1,14 +1,24 @@
-    "Congrats, you're up and running! Welcome to the first moshtix code challenge! " +
-    "Please replace this message with your solution and feel free to add " +
-    "or modify other files as necessary. Good luck : )";
+    for (counter = 1 ; counter <= 100; counter++) {
```
Are you aware of a plugin/library that could help keep code style consistent? No biggie, but I can see use of semi-colon and whitespace inconsistencies here.

homeindexjeff replied:
I normally code in visual studio to keep code styling consistent, although I must of went back and changed one of the values which interfered with the styling of that part.

danwfeeney commented on test-one/src/index.js:
```diff
@@ -1,14 +1,24 @@
-    "Please replace this message with your solution and feel free to add " +
-    "or modify other files as necessary. Good luck : )";
+    for (counter = 1 ; counter <= 100; counter++) {
+        if (counter % 3 && counter % 5) {
```
Might there be benefit in moving all this logic into it's own file so that it can be reused? Can you think of any other benefit in moving into it's own file?

homeindexjeff replied:
Another benefit would be letting it be able to be easier to read and understand.

danwfeeney commented on test-one/src/index.js:
```diff
@@ -1,14 +1,24 @@
-    "Please replace this message with your solution and feel free to add " +
-    "or modify other files as necessary. Good luck : )";
+    for (counter = 1 ; counter <= 100; counter++) {
+        if (counter % 3 && counter % 5) {
```
Can you think of another way of writing this line of code?

homeindexjeff replied:
Yes, I could of used some switch statement or while loops like below:

import { log } from "./helpers/logger";

export const go = async () => {
    var counter = 1;
    var displayMessage = "";

    while (counter <= 100) {
        switch (counter % 3 && counter % 5) {
            default: displayMessage += counter; break;
            case 0:
                if (counter % 3 == 0) {
                    displayMessage += "Boss";
                }
                if (counter % 5 == 0) {
                    displayMessage += "Hog";
                }; break;     
        }
        displayMessage += "\n";
        counter++
    }

    await log(displayMessage);
};

go();

danwfeeney commented on test-two/src/index.js:
```diff
@@ -4,23 +4,29 @@ import { getUsersForOrganisation } from "./services/github";
 const baseUrl = "https://api.github.com/";
 const organisationId = "google";
 const gitHubApiAuthToken = ""; // add your GitHub API OAuth key here to increase request limit
+var pageNumber;
```
Shouldn't we be using const or let?

homeindexjeff replied:
I thought const  was only used for variables that can't be reassigned and I didn't think let would of scooped properly.

danwfeeney commented on test-two/src/index.js:
```diff
@@ -4,23 +4,29 @@ import { getUsersForOrganisation } from "./services/github";
-      return `${result}Username: ${user.login}\n`;
-    }, "");
+    for (pageNumber = 1; pageNumber <= 23; pageNumber++) {
+        try {
```
Is this the ideal place for try catch in this file?

homeindexjeff replied:
No, I should of put the try and catch outside the for loop sorry.

danwfeeney commented on test-two/src/index.js:
```diff
@@ -4,23 +4,29 @@ import { getUsersForOrganisation } from "./services/github";
-    const output = users.reduce((result, user) => {
-      return `${result}Username: ${user.login}\n`;
-    }, "");
+    for (pageNumber = 1; pageNumber <= 23; pageNumber++) {
```
What will happen when another 100 users work for google?

danwfeeney replied:
Or if we wanted to change the organisation?

homeindexjeff replied:
Yeah, I was running out of time towards the end because my place had issues with the internet and I was pretty fatigue from a side effect of a vaccine I had. To improve the code it would of been better for me to of stored the list of names in an array and had the loop look for if it started pulling null response.

danwfeeney commented on test-two/src/index.js:
```diff
@@ -4,23 +4,29 @@ import { getUsersForOrganisation } from "./services/github";
-    const output = users.reduce((result, user) => {
-      return `${result}Username: ${user.login}\n`;
-    }, "");
+    for (pageNumber = 1; pageNumber <= 23; pageNumber++) {
```
Can you think of a way to improve the performance of this for loop?

homeindexjeff replied:
Yes, I could of moved the try and catch block out of the loop and had the results sorted in an array and have the them displayed after the loop was completely finished.

danwfeeney reviewed (commented):
Nice work 👍 You've met the requirements of each test, but I've left some feedback I'd like you to think about and respond to.
